### ISSUE TITLE
Add functionality for generatins a random string that matches a regex

### ISSUE THREAD

**Issue by GarkGarcia** [afrantzis/regen#1]
I would be cool if the CLI tool had an option to generate a random string matching the pattern specified in the command-line arguments

**Comment by GarkGarcia** [comment]
It doesn't seem that hard of feature to implement. If we where to make `Generator::append_next` generic, we could simply simply generate a random number `n` and write the first `n` strings to `/dev/null`. This is a bit of a stupid solution, but it would work.

**Comment by afrantzis** [comment]
This has been on my radar, too.

The issue with using a random number is that, in order for this to be "fair", that number should be selected uniformly from within the range of outputs. At the moment, the only way to get the number of possible outputs is to produce them. So we would need a first run to get the number of outputs (like the CLI already does, or even more efficiently with the dummy std::io::Write you propose) and a second run to get the one we want. This is certainly doable, but not very efficient, although possibly good/fast enough for many cases.

Another approach I have been considering is to be able to randomly set the "starting" state of the Generator (and thus AppendingIterators) so the next value produced is in effect random. Some complications may arise when the maximum output length is set, since the random state may lead to larger outputs. In that case one way out is just to retry until we get something valid.

**Comment by GarkGarcia** [comment]
> The issue with using a random number is that, in order for this to be "fair", that number should be selected uniformly from within the range of outputs. At the moment, the only way to get the number of possible outputs is to produce them. So we would need a first run to get the number of outputs (like the CLI already does, or even more efficiently with the dummy std::io::Write you propose) and a second run to get the one we want. This is certainly doable, but not very efficient, although possibly good/fast enough for many cases.

I believe we could mathematically derive the number of possible matches,  couldn't we? For example, the number of possible matches for `[<a group of n character>]{<a natural number m>}` is simply `n^m`. That even extends to cases where there are infinitely many possible matches. For example, the number of possible matches for `<any character group>{<any natural number>,}` is infinite.

Composing simple rules, like the ones I described, recursively should give us the total number of matches (finite and infinite) without the need to compute them all. It wouldn't be the easiest thing to implement right now, but it seems quite possible.

The would vastly improve the efficiency of the _count_ option in the CLI too.

> Another approach I have been considering is to be able to randomly set the "starting" state of the Generator (and thus AppendingIterators) so the next value produced is in effect random. Some complications may arise when the maximum output length is set, since the random state may lead to larger outputs. In that case one way out is just to retry until we get something valid.

The maximum length thing is definitively an issue in both the mathematical approach and the approach suggested in your last paragraph. I think I could ignore it for now, and try to implement it later (?). Not sure about this one.

**Comment by GarkGarcia** [comment]
Also worth noting: the number of possible matches to `(<some group "a">)(<some other group "b">)` is `n * m`, where `n` is the number of possible matches to `a` and `m` is number of possible matches to `b`.

**Comment by afrantzis** [comment]
> Composing simple rules, like the ones I described, recursively should give us the total number of matches (finite and infinite) without the need to compute them all. It wouldn't be the easiest thing to implement right now, but it seems quite possible.

For unbounded outputs the count can indeed be calculated using simple combinatorics formulas. I was previously hesitant to introduce a method e.g., Generator::count(), which works only for some cases, but perhaps it's just a matter of framing expectations properly. For example, it could be something like Generator::try_quick_count() -> Option<usize>, which I think sets the expectations clearly enough.

> For example, the number of possible matches for <any character group>{<any natural number>,} is infinite.

This got me thinking about what makes sense as a properly random string in the context of infinite options. I can't find any sense in that, so perhaps getting a random string should only be allowed when a maximum length is implicitly (i.e., the regex itself is self limiting) or explicitly set.

**Comment by GarkGarcia** [comment]
> For unbounded outputs the count can indeed be calculated using simple combinatorics formulas. I was previously hesitant to introduce a method e.g., Generator::count(), which works only for some cases, but perhaps it's just a matter of framing expectations properly. For example, it could be something like Generator::try_quick_count() -> Option, which I think sets the expectations clearly enough.

Keep in mind this could be an internal method too, that way user expectations aren't that much of a concert. Maybe we could name the method `Generator::count_hint` and make it return `min(n, m)`, where `n` is the number of possible matches of the pattern (finite or infinite) and `m` is the number of strings that fit the character bound (strings whose length is less than or equals to the user defined maximum number of character).

**Comment by GarkGarcia** [comment]
> This got me thinking about what makes sense as a properly random string in the context of infinite options. I can't find any sense in that, so perhaps getting a random string should only be allowed when a maximum length is implicitly (i.e., the regex itself is self limiting) or explicitly set.

Indeed, I wasn't able to find much sense in it too. However, I don't think most users are interested is encryption-grade randomness or anything similar. Replacing, `infinity` with `usize::max()` or something similar should work well enough. We could also parameterize the method so that the caller is required to pass the maximum character count. This would require some refactoring thought, since currently the maximum character count is baked into the struct itself.
